### ISSUE TITLE
page_service doesn't handle messages split across TCP packets correctly

### ISSUE THREAD

**Issue by hlinnaka** [zenithdb/zenith#66]
I noticed while reading the code that `FeMessage::parse` function consumes the bytes it reads from the buffer, even if the whole message is not present. That causes trouble if the message is split across multiple TCP packets. To reproduce, I created a proxy that forwards the data one byte at a time, i.e. each byte is sent as separate TCP packet:
```
# start pageserver, listening on port 64000
./target/debug/pageserver -l 127.0.0.1:64000

# in another terminal, create proxy:
socat -b1 TCP-LISTEN:64002,reuseaddr TCP:localhost:64000

# connect with psql:
psql -p 64002 -h localhost

# pageserver panics:
thread 'tokio-runtime-worker' panicked at 'cannot advance past `remaining`: 80 <= 8', /home/heikki/.cargo/registry/src/github.com-1ecc6299db9ec823/bytes-1.0.1/src/bytes_mut.rs:953:9
```
If you replace `-b1` above with a higher number, it works.

**Comment by ericseppanen** [comment]
Can this be handled with a `BufReader`?

**Comment by hlinnaka** [comment]
Fixed by commit 3b9e7fc5e6715a694b27e059ee7eedd1f6f44735

**Comment by hlinnaka** [comment]
> Can this be handled with a `BufReader`?

BufReader only changes the performance characteristics of the underlying `std::io::Read` I believe, so no. I used a BufReader anyway though, for the performance reasons, and for symmetry with the output side of the stream which already used BufWriter. But this got fixed by the other refctorings in 3b9e7fc5e6715a694b27e059ee7eedd1f6f44735
